examples/cis2-multi-royalties/src/lib.rs

```diff
@@ -0,0 +1,1405 @@
+}
+
+/// View function for testing. This reports on the entire state of the contract
+/// for testing purposes. In a realistic example there `balance_of` and similar
+/// functions with a smaller response.
+#[receive(contract = "cis2_multi_royalties", name = "view", return_value = "ViewState")]
+fn contract_view<S: HasStateApi>(
+    _ctx: &impl HasReceiveContext,
+    host: &impl HasHost<State<S>, StateApiType = S>,
+) -> ReceiveResult<ViewState> {
+    let state = host.state();
+
+    let mut inner_state = Vec::new();
+    for (k, a_state) in state.state.iter() {
+        let mut balances = Vec::new();
+        let mut operators = Vec::new();
+        for (token_id, amount) in a_state.balances.iter() {
+            balances.push((*token_id, *amount));
+        }
+        for o in a_state.operators.iter() {
+            operators.push(*o);
+        }
+
+        inner_state.push((*k, ViewAddressState {
+            balances,
+            operators,
+        }));
+    }
+    let mut tokens = Vec::new();
+    for v in state.tokens.iter() {
+        tokens.push(*v);
+    }
+
+    Ok(ViewState {
+        state: inner_state,
+        tokens,
+    })
+}
+
+/// Mint new tokens with a given address as the owner of these tokens.
+/// Can only be called by the contract owner.
+/// Logs a `Mint` and a `TokenMetadata` event for each token.
+/// The url for the token metadata is the token ID encoded in hex, appended on
+/// the `TOKEN_METADATA_BASE_URL`.
+///
+/// It rejects if:
+/// - The sender is not the contract instance owner.
+/// - Fails to parse parameter.
+/// - Any of the tokens fails to be minted, which could be if:
+///     - Fails to log Mint event.
+///     - Fails to log TokenMetadata event.
+///
+/// Note: Can at most mint 32 token types in one call due to the limit on the
+/// number of logs a smart contract can produce on each function call.
+#[receive(
+    contract = "cis2_multi_royalties",
+    name = "mint",
+    parameter = "MintParams",
+    error = "ContractError",
+    enable_logger,
+    mutable
+)]
+fn contract_mint<S: HasStateApi>(
+    ctx: &impl HasReceiveContext,
+    host: &mut impl HasHost<State<S>, StateApiType = S>,
+    logger: &mut impl HasLogger,
+) -> ContractResult<()> {
+    // Get the contract owner
+    let owner = ctx.owner();
+    // Get the sender of the transaction
+    let sender = ctx.sender();
+
+    ensure!(sender.matches_account(&owner), ContractError::Unauthorized);
+
+    // Parse the parameter.
+    let params: MintParams = ctx.parameter_cursor().get()?;
+
+    let (state, builder) = host.state_and_builder();
+    for (token_id, token_amount) in params.tokens {
+        // Mint the token in the state.
+        state.mint(&token_id, token_amount, &params.owner, builder, 0);
+        
+        // Event for minted token.
+        logger.log(&Cis2Event::Mint(MintEvent {
+            token_id,
+            amount: token_amount,
+            owner: params.owner,
+        }))?;
+
```

Comment:
Consider adding the `royalty` to the event. The percentage value of how much `royalties` are paid might be interesting to  off-chain tool.